share/model.go

```diff
@@ -30,34 +30,55 @@ func (i ItemType) String() string {
 	return string(i)
 }
 
+// ShareID only contains IDs of shares and public links. This is because OCIS requires
+// that shares and public links do not share an ID, so we need a shared table to make sure
+// that there are no duplicates.
+// This is implemented by having ShareID have an ID that is auto-increment, and shares and
+// public links will have their ID be a foreign key to ShareID
+// When creating a new share, we will then first create an ID entry and use this for the ID
+
+type ShareID struct {
+	ID uint `gorm:"primarykey"`
+}
+
+// We cannot use gorm.Model, because we want our ID to be a foreign key to ShareID
+type BaseModel struct {
+	// Id has to be called Id and not ID, otherwise the foreign key will not work
+	// ID is a special field in GORM, which it uses as the default Primary Key
+	Id        uint    `gorm:"uniqueIndex"`
+	ShareId   ShareID `gorm:"foreignKey:Id;references:ID;constraint:OnDelete:CASCADE"` //;references:ID
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	DeletedAt gorm.DeletedAt `gorm:"index"`
+}
+
 // ProtoShare contains fields that are shared between PublicLinks and Shares.
 // Unfortunately, because these are shared, we cannot name our indexes
 // because then two indexes with the same name would be created
 type ProtoShare struct {
 	// Including gorm.Model will embed a number of gorm-default fields
-	gorm.Model
+	BaseModel
 	UIDOwner     string   `gorm:"size:64"`
```

Comment:
Should we also have an index here? Our queries are basically "what have I/someone shared" and "with whom have I/someone shared with". Any implication?

Reply:
On `UIDInitiator` then :) but yes, good point!